napari/utils/_magicgui.py

```diff
@@ -16,21 +16,132 @@
 
 if TYPE_CHECKING:
     from concurrent.futures import Future
 
-    from magicgui.widgets import FunctionGui
-    from magicgui.widgets._bases import CategoricalWidget
+    from magicgui.widgets.bases import CategoricalWidget
 
     from napari._qt.qthreading import FunctionWorker
     from napari.layers import Layer
     from napari.viewer import Viewer
 
 
+def _get_layer_from_widget(gui: ComboBox, viewer: Viewer) -> Optional[Layer]:
+    """Retrieve layer used as input to function wrapped into magicgui .
+
+    Parameters
+    ----------
+    gui : magicgui.widgets.ComboBox
+        The instantiated ComboBox widget.
+
+    Returns
+    -------
+    Layer
+        layer passed as input to the function wrapped into magicgui.
+
+    """
+    from napari.layers import Layer
+
+    if isinstance(gui.value, Layer):
+        return gui.value
+
+    if gui.annotation is None or gui.annotation.__module__ != 'napari.types':
+        return None
+
+    layer_name = gui.current_choice[: -len(' (data)')]
+    return viewer.layers[layer_name]
+
+
+def _get_layers_from_widget(gui: FunctionGui, viewer: Viewer) -> list[Layer]:
+    """Retrieve layers used as input to function wrapped into magicgui .
+
+    Parameters
+    ----------
+    gui : magicgui.widgets.FunctionGui
+        The instantiated FunctionGui widget.
+
+    Returns
+    -------
+    list
+        list of layers passed as input to the function wrapped into magicgui.
+
+    """
+    res = []
+    for widg in gui:
+        if isinstance(widg, ComboBox):
+            layer = _get_layer_from_widget(widg, viewer)
+            if layer is not None:
+                res.append(layer)
+    return res
+
+
+def _get_spatial_from_layer(layer, ndim):
+    axis = np.array(range(-ndim, 0))
+
+    data2physical = layer._transforms['data2physical'].set_slice(axis)
+    return {
+        'scale': data2physical.scale,
+        'translate': data2physical.translate,
+        'rotate': data2physical.rotate,
+        'shear': data2physical.shear,
+        'affine': layer.affine.set_slice(axis).affine_matrix,
+    }
+
+
+def _cmp_meta(meta1, meta2):
+    return (
+        np.array_equal(meta1['scale'], meta2['scale'])
+        and np.array_equal(meta1['translate'], meta2['translate'])
+        and np.array_equal(meta1['rotate'], meta2['rotate'])
+        and np.array_equal(meta1['shear'], meta2['shear'])
+        and np.array_equal(meta1['affine'], meta2['affine'])
+    )
+
+
+def _calc_affine_from_source_layers(data, source_layers: list[Layer]):
+    """Calculate affine information for provided data based on source layers.
+
+    Parameters
+    ----------
+    data : array
+        Data to be added to the viewer.
+    source_layers : list
+        List of layers used to calculate affine information.
+
+    Returns
+    -------
+    dict
+        Dictionary with affine information.
+    """
+
+    source_layers = [
+        layer for layer in source_layers if layer.ndim >= data.ndim
```

Comment:
Why is it ok to ignore the lower dimensionality layers? Should we also show the warning in this case? E.g. if you have a 2D layer with scale `(3, 3)` and you add a 3D stack, it's quite likely user would want scale on the 3D layer to be `(1, 3, 3)`. I don't think we should guess this, but it should probably lead to the same warning as below.

Reply:
Because guessing is hard. I will add warning.